web3_proxy/src/stats/mod.rs

```diff
@@ -301,12 +323,212 @@ impl BufferedRpcQueryStats {
+            .await?;
+
+        // Technicall there should always be a user ... still let's return "Ok(())" for now
+        let sender_user_id: u64 = match sender_rpc_key {
+            Some(x) => x.user_id.into(),
+            // Return early if the User is not found, because then it is an anonymous user
+            // Let's also issue a warning because obviously the RPC key should correspond to a user
+            None => {
+                warn!(
+                    "No user was found for the following rpc key: {:?}",
+                    rpc_secret_key_id
+                );
+                return Ok(());
+            }
+        };
+
+        // (1) Do some general bookkeeping on the user
+        let sender_balance = match balance::Entity::find()
+            .filter(balance::Column::UserId.eq(sender_user_id))
+            .one(db_conn)
+            .await?
+        {
+            Some(x) => x,
+            None => {
+                warn!("This user id has no balance entry! {:?}", sender_user_id);
+                return Ok(());
+            }
+        };
+
+        let mut active_sender_balance = sender_balance.clone().into_active_model();
+
+        // Still subtract from the user in any case,
+        // Modify the balance of the sender completely (in mysql, next to the stats)
+        // In any case, add this to "spent"
+        active_sender_balance.used_balance =
+            sea_orm::Set(sender_balance.used_balance + Decimal::from(self.sum_credits_used));
+
+        // Also update the available balance
+        let new_available_balance = max(
+            sender_balance.available_balance - Decimal::from(self.sum_credits_used),
+            Decimal::from(0),
+        );
+        active_sender_balance.available_balance = sea_orm::Set(new_available_balance);
+
+        active_sender_balance.save(db_conn).await?;
+
+        let downgrade_user = match user::Entity::find()
+            .filter(user::Column::Id.eq(sender_user_id))
+            .one(db_conn)
+            .await?
+        {
+            Some(x) => x,
+            None => {
+                warn!("No user was found with this sender id!");
+                return Ok(());
+            }
+        };
+
+        let downgrade_user_role = user_tier::Entity::find()
+            .filter(user_tier::Column::Id.eq(downgrade_user.user_tier_id))
+            .one(db_conn)
+            .await?
+            .context(format!(
+                "The foreign key for the user's user_tier_id was not found! {:?}",
+                downgrade_user.user_tier_id
+            ))?;
+
+        // Downgrade a user to premium - out of funds if there's less than 10$ in the account, and if the user was premium before
+        if new_available_balance < Decimal::from(10u64) && downgrade_user_role.title == "Premium" {
+            let mut active_downgrade_user = downgrade_user.into_active_model();
+            active_downgrade_user.user_tier_id = sea_orm::Set(downgrade_user_role.id);
```

Comment:
Wait. Do we actually change the role in the db? I don't think we want that. It's likely we will have multiple roles that all downgrade to the same place. But then we won't know how to restore their roll when they deposit new funds. 

Instead, when we fetch the cached user data, we should set the user tier id there to be the downgraded if their balance is 0